### ISSUE TITLE
SpartaStaticInitializer initialization fails "Attempted to access StringManager singleton before it was statically constructed"

### ISSUE THREAD

**Issue by nilanjan** [sparcians/map#171]
StaticInitializer fails for the following lines in `StringManager` singleton `getStringManger()` function:
```
        static StringManager& getStringManager() {
            sparta_assert(_GBL_string_manager && _GBL_string_manager->is_constructed == IS_CONSTRUCTED_CONST,
                              "Attempted to access StringManager singleton before it was "
                              "statically constructed.")
            return *_GBL_string_manager;
        }
```
 due to `SpartaStaticInitializer()` constructor in sparta.cpp not properly assigning the value. Even after this statement `_GBL_string_manager` is `NULL`.
```
StringManager::_GBL_string_manager = new StringManager();
```
After some investigation found that the construction of `StringManager()` is modified to have a setter method. This is the hack that worked. 
`StringManager::setStringManager(new StringManager());`

This kind of issues are visible for most of the global variables. 

I am now facing this for `KeyValue::GBL_type_name_map` in the following line of code within macro `SPARTA_KVPAIR_BODY`:
`KeyValue::GBL_type_name_map[typeid(type).name()] = boost::trim_left_copy(*current_type_);`

**OS**
Mac OS Catlina
**Compiler**
clang version 10.0.0 
Target: x86_64-apple-darwin19.5.0
Thread model: posix
Cmake build as dependent project. 
**Build/Make strategy**
Multiple inclusion of the libsparta.dylib into different intermediate libraries and then all of those intermediate libraries including libsparta.dylib are linked to the final executable.

**Comment by knute-sifive** [comment]
We've never tested Sparta as a shared library.  It obviously won't work due to the rules of shared library and static variables.  Can you not build Sparta as a static library?

**Comment by nilanjan** [comment]
Thanks @knute-sifive . I have converted all the libraries to static and it ran successfully.

**Comment by nilanjan** [comment]
Issue resolved if all the libraries are static.

**Comment by knute-sifive** [comment]
OOC, what was your `cmake` configure line to enable this?  I'd like to at least add an error condition to prevent folks from running into this problem again.

**Comment by nilanjan** [comment]
I was manually modifying the sparta/CMakeLists.txt to generate SHARED library. And all the other internal libraries were SHARED. As an intermediate step I make sparta STATIC library and rest of the project SHARED. This was still showing the same issue. Finally, I converted all the internal libraries STATIC including sparta. Then it ran successfully.

**Comment by jackjchen** [comment]
@nilanjan would you mind sharing your final CMakeLists.txt to a https://gist.github.com/ Thanks.

**Comment by nilanjan** [comment]
It will be hard for me to share the entire cmake file. But I will post some excerpts from the cmake file.

**Comment by nilanjan** [comment]
Most of the add_library() is converted to generate SHARED lib within sparta and my project all the add_library() were set to SHARED. Thats basically what I did. Once I moved to STATIC mode the advantage is I do not need to modify anything on sparta lib. I am able to use TOT unchanged. My suggestion would be to mention in the SPARTA readme that use the library as STATIC; do not try to generate dynamic library. Later I may try to support dynamic if you think there no fundamental problem that would restrict me to do so. As of now the generated binary is not too big for me to really bother about it.
